[aot_autograd] only performance functionalization analysis pass once

bdhirsh opened the pull request:
For a while now, we've been re-running our functionalization analysis pass twice - once for get metadata when dedup'ing, and an entire second time during aot_dispatch_base/autograd.

This should also probably speed up compile times pretty noticeably, since we're going from:

(a) inference-only trace case: 3 fw traces -> 2 fw traces
(b) autograd trace case: 2 fw traces + 1 joint trace -> 1 fw trace + 1 joint trace


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #96784
* #96342
* #96341
* #96340
* #96339
* #96235
* __->__ #95992



cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

bdhirsh committed:
[aot_autograd] only performance functionalization analysis pass once

[ghstack-poisoned]

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1812,7 +1794,7 @@ def add_dupe_args(args):
         return flat_fn(*add_dupe_args(args))
 
-    compiled_fn = compiler_fn(wrapped_flat_fn, deduped_flat_args, aot_config)
+    compiled_fn = compiler_fn(wrapped_flat_fn, deduped_flat_args, aot_config, fw_metadata=fw_metadata)
```
If you deduped the args, don't you have to update the `fw_metadata`?

bdhirsh replied:
hmmm good catch.

I still want to avoid computing the metadata twice. There are probably a few ways of dealing with this, but here's what I'm thinking: `aot_dedupe()` will take in a metadata object that includes the analysis on dupe'd args, and only if necessary, return a new metadata object that is the same but with input metadata dedup'd.

ezyang replied:
Yeah, having the deduper also dedupe the metadata soudns reasonable (if fiddly)

bdhirsh replied:
this is done

bdhirsh committed:
Update on "[aot_autograd] only performance functionalization analysis pass once"

For a while now, we've been re-running our functionalization analysis pass twice - once for get metadata when dedup'ing, and an entire second time during aot_dispatch_base/autograd.

This should also probably speed up compile times pretty noticeably, since we're going from:

(a) inference-only trace case: 3 fw traces -> 2 fw traces
(b) autograd trace case: 2 fw traces + 1 joint trace -> 1 fw trace + 1 joint trace





[ghstack-poisoned]

bdhirsh committed:
Update on "[aot_autograd] only performance functionalization analysis pass once"

For a while now, we've been re-running our functionalization analysis pass twice - once for get metadata when dedup'ing, and an entire second time during aot_dispatch_base/autograd.

This should also probably speed up compile times pretty noticeably, since we're going from:

(a) inference-only trace case: 3 fw traces -> 2 fw traces
(b) autograd trace case: 2 fw traces + 1 joint trace -> 1 fw trace + 1 joint trace





[ghstack-poisoned]

bdhirsh committed:
Update on "[aot_autograd] only performance functionalization analysis pass once"

For a while now, we've been re-running our functionalization analysis pass twice - once for get metadata when dedup'ing, and an entire second time during aot_dispatch_base/autograd.

This should also probably speed up compile times pretty noticeably, since we're going from:

(a) inference-only trace case: 3 fw traces -> 2 fw traces
(b) autograd trace case: 2 fw traces + 1 joint trace -> 1 fw trace + 1 joint trace





[ghstack-poisoned]

ezyang commented on torch/_functorch/aot_autograd.py:
```diff
@@ -1595,6 +1587,37 @@ def format_guard_bug_msg(aot_config, expected):
+            for o in m.output_info],
+        num_intermediate_bases=m.num_intermediate_bases,
+        keep_input_mutations=m.keep_input_mutations,
+        traced_tangents=m.traced_tangents,
```
As a debug option, it would be nice to be able rerun the analysis from scratch and compare that the manually deduped metadata agrees with the manual analysis.

bdhirsh replied:
yep definitely a good idea

bdhirsh replied:
debug asserts are currently on by default - it might make sense to turn them off by default now (and only turn them on by default for the aot_eager backend?). In my next PR, I do the same thing with synthetic bases - so debug asserts would mean we trace the forward with functionalization 2 extra times.

bdhirsh replied:
(for now I'm making that change, to only turn them on in our test_aotdispatch tests + when using aot_eager)

bdhirsh committed:
Update on "[aot_autograd] only performance functionalization analysis pass once"

For a while now, we've been re-running our functionalization analysis pass twice - once for get metadata when dedup'ing, and an entire second time during aot_dispatch_base/autograd.

This should also probably speed up compile times pretty noticeably, since we're going from:

(a) inference-only trace case: 3 fw traces -> 2 fw traces
(b) autograd trace case: 2 fw traces + 1 joint trace -> 1 fw trace + 1 joint trace





[ghstack-poisoned]

bdhirsh committed:
Update on "[aot_autograd] only performance functionalization analysis pass once"

For a while now, we've been re-running our functionalization analysis pass twice - once for get metadata when dedup'ing, and an entire second time during aot_dispatch_base/autograd.

This should also probably speed up compile times pretty noticeably, since we're going from:

(a) inference-only trace case: 3 fw traces -> 2 fw traces
(b) autograd trace case: 2 fw traces + 1 joint trace -> 1 fw trace + 1 joint trace





cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "[aot_autograd] only performance functionalization analysis pass once"

For a while now, we've been re-running our functionalization analysis pass twice - once for get metadata when dedup'ing, and an entire second time during aot_dispatch_base/autograd.

This should also probably speed up compile times pretty noticeably, since we're going from:

(a) inference-only trace case: 3 fw traces -> 2 fw traces
(b) autograd trace case: 2 fw traces + 1 joint trace -> 1 fw trace + 1 joint trace





cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]
